OpenRA.Mods.Common/SpriteLoaders/GifLoader.cs

```diff
@@ -0,0 +1,311 @@
+								ReadSubblock(s);
+								break;
+							case BlockTypes.PlainTextExtension:
+								var lengthP = s.ReadUInt8();
+								s.Position += lengthP;
+								ReadSubblock(s);
+								break;
+							case BlockTypes.ApllicationExtensionBlock:
+								var lengthA = s.ReadUInt8();
+								if (lengthA != 11)
+									throw new Exception("Bad app extension");
+
+								var identifier = s.ReadASCII(8);
+								s.Position += 3;
+								if (identifier == "NETSCAPE")
+									s.Position += 5;
+								else
+									ReadSubblock(s);
+								break;
+							default:
+								ReadSubblock(s);
+								break;
+						}
+						break;
+					case BlockTypes.ImageDescriptor:
+						frames.Add(new GifFrame(s, size));
+						break;
+					case BlockTypes.Trailer:
+						return frames.ToArray();
+					default:
+						Console.WriteLine("Bad data");
+						break;
+				}
+
+				// Shouldn't happen but if Trailer is missing at EOF we should bail.
+				if (s.Position >= s.Length)
+					break;
+			}
+
+			return frames.ToArray();
+		}
+
+		public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
+		{
+			if (!IsGif(s))
+			{
+				frames = null;
+				return false;
+			}
+
+			frames = ParseFrames(s);
+			return true;
+		}
+
+		#region Helper functions
+
+		static int BitsToNum(byte[] array)
+		{
+			var value = (int)(array[0]);
+			for (var i = 1; i < array.Length; i++)
+				value = value * 2 + array[i];
+
+			return value;
+		}
+
+		// Each entry is 3 bytes in RGB
+		static byte[] ParseColorTable(Stream s, int entries)
+		{
+			var ct = new List<byte>();
+			for (var i = 0; i < entries; i++)
+				ct.AddRange(s.ReadBytes(3));
+
+			return ct.ToArray();
+		}
+
+		static byte[] ReadSubblock(Stream s)
+		{
+			var data = new List<byte>();
+
+			while (true)
+			{
+				var ssize = s.ReadUInt8();
+				if (ssize == 0)
+					break;
+
+				data.AddRange(s.ReadBytes(ssize));
+			}
+
+			return data.ToArray();
+		}
+
+		static byte[] Decode(int minCodeSize, byte[] data)
+		{
+			var pos = 0;
+			var code = 0;
+			var last = 0;
+
+			var decrypt = new Action<int>((size) =>
+			{
+				code = 0;
+				for (var i = 0; i < size; i++)
+				{
+					if ((data[pos >> 3] & (1 << (pos & 7))) != 0)
+						code |= 1 << i;
+					pos++;
+				}
+			});
+
+			var output = new List<byte>();
+
+			var clearCode = 1 << minCodeSize;
+			var eoiCode = clearCode + 1;
+
+			var codeSize = minCodeSize + 1;
+
+			var dict = new List<byte[]>();
+
+			while (true)
+			{
+				last = code;
+				decrypt(codeSize);
+
+				if (code == clearCode)
+				{
+					dict.Clear();
+					codeSize = minCodeSize + 1;
+					for (var i = 0; i < clearCode; i++)
+						dict.Add(new byte[]{(byte)i});
+					dict.Add(new byte[]{});
+					dict.Add(null);
+					continue;
+				}
+				
+				if (code == eoiCode)
+					break;
+
+				if (code < dict.Count)
+					if (last != clearCode)
+						dict.Add(dict.ElementAt(last).Concat(new byte[]{dict.ElementAt(code)[0]}).ToArray());
+				else
```

Comment:
This is indented differently from how the compiler will see this code, this else will be matched with the nested if and not the outer if.

Either you have a severe logic error or the indentation is off. I recommend some braces.